sktime/pipeline/_graph_pipe.py

```diff
@@ -0,0 +1,210 @@
+    - steps : list of GraphNode objects, (optional), default = None,
+        contains the GraphNode objects in the order they should be fit.
+
+    Attributes
+    ----------
+
+    - steps : list,
+    - assembled_steps : list,
+    - outputs : GraphOutput,
+
+    Examples
+    --------
+    """
+
+    parameters = {
+        "steps": list,
+        "assembled_steps": list,
+        "estimators": list,
+        "outputs": GraphOutput,
+    }
+
+    @classmethod
+    def assemble(cls, steps: list, outputs=None):
+        estimators = [node.get_estimator() for node in steps]
+        if not outputs:
+            outputs = GraphOutput()
+        est_dict = {}
+        for est in estimators:
+            est_dict[id(est)] = clone(est)
+        assembled_steps = []
+        for node in steps:
+            new_node = GraphNode(
+                est_dict[id(node.estimator)],
+                node.name,
+                node.tags,
+                node.input_edges,
+                node.input_indices,
+            )
+            assembled_steps.append(new_node)
+        params = {
+            "steps": steps,
+            "assembled_steps": assembled_steps,
+            "estimators": estimators,
+            "outputs": outputs,
+        }
+
+    def __init__(
+        self,
+        steps: list = [],
+    ):
+        params = GraphPipeline.assemble(steps)
+        self.set_params(**params)
+
+    def add(
+        self, name: str, estimator: BaseEstimator, input_edges, input_indices, tags
+    ):
+        """add a new node to the end of the GraphPipeline"""
+        new_node = GraphNode(
+            estimator,
+            name,
+            tags,
+            input_edges,
+            input_indices,
+        )
+        self.steps.append(new_node)
+        params = GraphPipeline.assemble(self.steps)
+        self.set_params(**params)
+
+    def fit(self, y, X=None):
+        """Actually fit the GraphPipeline:"""
+
+        self.outputs["raw_input"] = (y, X)
+        for node in self.assembled_steps:
+            input_y, input_X = self.outputs.get_input(
+                node.input_edges, node.input_indices
+            )
+            if isinstance(node.get_estimator, BaseTransformer):
+                transform_args = {"X": input_X, "y": input_y}
+                if "y" in node.tags:  # if y in in transform, switch order of inputs
+                    transform_args = {"X": input_y, "y": input_X}
+                if "inverse" in node.tags:
+                    output = node.estimator.inverse_transform(**transform_args)
+                else:
+                    output = node.estimator.fit_transform(**transform_args)
+                if "y" in node.tags:
+                    input_y = output
+                else:
+                    input_X = output
+            if isinstance(node.get_estimator, BaseForecaster):
+                node.estimator.fit(y=input_y, X=input_X)
+                input_y, input_X = node.estimator.y, node.estimator.X
```

Comment:
This can be probably extracted into the steps or graphnodes (we have to decide on the naming graphnode vs step). Then this could be solved via inheritance. On specific graphnode per base class.